cmd: Fix issue reading string map type via config map

sayboras opened the pull request:
### Description

As mentioned in below upstream issue, there is a discrepancy in viper
while reading string map string data type i.e. kv pair format was not
supported, only `{"k":"v"}` format is allowed. This commit is to wrap
GetStringMapString implementation to handle such case. This is backward
compatible change.

Relates https://github.com/spf13/viper/issues/911
Fixes https://github.com/cilium/cilium/issues/18328

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

### Testing

Testing was done locally as per below

<details>
<summary>before</summary>

```
$ cat tmp/subnet-tags-filter
k1=v1,k2=v2% 

/usr/lib/go-1.17/bin/go build -o /tmp/GoLand/___2go_build_github_com_cilium_cilium_operator github.com/cilium/cilium/operator #gosetup
/tmp/GoLand/___2go_build_github_com_cilium_cilium_operator --config-dir tmp/
level=info msg="Started gops server" address="127.0.0.1:9891" subsys=___2go_build_github_com_cilium_cilium_operator
level=warning msg="Running Cilium with \"kvstore\"=\"\" requires identity allocation via CRDs. Changing identity-allocation-mode to \"crd\"" subsys=config
...
level=info msg="  --subnet-ids-filter=''" subsys=___2go_build_github_com_cilium_cilium_operator
level=info msg="  --subnet-ids-filter=''" subsys=___2go_build_github_com_cilium_cilium_operator --> THIS IS EMPTY
level=info msg="  --synchronize-k8s-nodes='true'" subsys=___2go_build_github_com_cilium_cilium_operator
...
level=info msg="Cilium Operator  go version go1.17 linux/amd64" subsys=___2go_build_github_com_cilium_cilium_operator
```

</details>


<details>
<summary>after</summary>

```
/usr/lib/go-1.17/bin/go build -o /tmp/GoLand/___2go_build_github_com_cilium_cilium_operator github.com/cilium/cilium/operator #gosetup
/tmp/GoLand/___2go_build_github_com_cilium_cilium_operator --config-dir tmp/
level=info msg="Started gops server" address="127.0.0.1:9891" subsys=___2go_build_github_com_cilium_cilium_operator
level=warning msg="Running Cilium with \"kvstore\"=\"\" requires identity allocation via CRDs. Changing identity-allocation-mode to \"crd\"" subsys=config

...
level=info msg="  --subnet-ids-filter=''" subsys=___2go_build_github_com_cilium_cilium_operator
level=info msg="  --subnet-tags-filter='k1=v1,k2=v2'" subsys=___2go_build_github_com_cilium_cilium_operator --> THIS HAS CORRECT VALUE
level=info msg="  --synchronize-k8s-nodes='true'" subsys=___2go_build_github_com_cilium_cilium_operator
...
level=info msg="Cilium Operator  go version go1.17 linux/amd64" subsys=___2go_build_github_com_cilium_cilium_operator
```

</details>


```release-note
cmd: Fix issue reading string map type via config map
```

sayboras committed:
cmd: Fix issue reading string map type via config map

As mentioned in below upstream issue, there is a discrepancy in viper
while reading string map string data type i.e. kv pair format was not
supported, only `{"k":"v"}` format is allowed. This commit is to wrap
GetStringMapString implementation to handle such case.

Also, during the bootstrap, if there is any flag having invalid value,
fatal log will be printed for early detection and awareness.

Relates https://github.com/spf13/viper/issues/911
Fixes https://github.com/cilium/cilium/issues/18328

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

sayboras commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package command
```
I look around, and find this package is the most relevant, feel free to let me know otherwise.

aanm replied:
We also have `tools/`, what do you think?

sayboras replied:
tools directory is used for extra utility (e.g. lint check, dev-doctor, etc), should not be used in any main application code.

sayboras commented on contrib/scripts/check-viper-get-string-map-string.sh:
```diff
@@ -0,0 +1,21 @@
+
+# Simple script to make sure viper.GetStringMapString should not be used.
+if grep -r --exclude-dir={.git,_build,vendor,contrib} -i --include \*.go "viper.GetStringMapString" .; then
+  echo "Found viper.GetStringMapString(key) usage. Please use command.GetStringMapString(viper.GetViper(), key) instead";
```
very naive check here, this will not work with import alias, but we all get used to viper.<something> I reckon.

glibsm commented on pkg/command/map_string_test.go:
```diff
@@ -0,0 +1,81 @@
+				value: `{"k1":"v1","k2":"v2",}`,
+			},
+			want: map[string]string{},
+		},
```
in the error case like this i would usually check the error message for `strings.Contains` something specific like `bad json` or `extra comma`, I'm not sure what sort of a message is printed here.

sayboras replied:
good point, I see the actual error in debugger only, making sense to make it explicit so that others don't need to re-check it. Let me refactor a little bit, and address related comments below as well.

glibsm commented on pkg/command/map_string_test.go:
```diff
@@ -0,0 +1,81 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetStringMapString(t *testing.T) {
```
some more diversity in input and output values would probably be welcome here, for instance `k1=v2, k1=v1`. What's the result there? I'd argue we should maybe error out

sayboras replied:
good point :100:, I actually miss this scenario in first implementation, thanks :100:

glibsm commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,43 @@
+		var syntaxErr *json.SyntaxError
+		if errors.As(err, &syntaxErr) {
+			switch s := data.(type) {
+			case string:
```
I think there is some potential of malformed json to not mix well with a value `=` here? Perhaps covering something like `{"foo": "bar",=asdf` would put me more at ease

sayboras replied:
thanks, I have added a test case for the same.

sayboras commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,85 @@
+	"github.com/spf13/viper"
+)
+
+var keyValueRegex = regexp.MustCompile(`((\w+)=(\w+,))+((\w+)=(\w+))$`)
```
for better visualisation and checking 

https://jex.im/regulex/#!flags=&re=((%5Cw%2B)%3D(%5Cw%2B%2C))%2B((%5Cw%2B)%3D(%5Cw%2B))%24

gandro replied:
Should this be `((\w+)=(\w+,))*((\w+)=(\w+))$` (note: `*` instead of `+` )? 

Otherwise, this will reject single `k=v` pairs, and always require at least two pairs (`k1=v1,k2=v2`).

sayboras replied:
Look at this the below unit test :face_exhaling: 

```
		{
			name: "valid format with one pair",
			args: args{
				str: "k1=v1,k2=v2", --> this should be only one kv pair as per description
			},
			want: true,
		},
``

sayboras commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,85 @@
+		case string:
+			// if the input is starting with either '{' or '[', just preserve original json parsing error.
+			firstIndex := firstNonSpaceCharIndex(s)
+			if firstIndex != -1 && (s[firstIndex] == '{' || s[firstIndex] == '[') {
```
the std json library didn't have dedicate error code for beginning of top level object, so checking first non space char here to make sure right error is returned.

sayboras commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,85 @@
+			kvs := strings.Split(s, ",")
+			for _, kv := range kvs {
+				temp := strings.Split(kv, "=")
+				if len(temp) != 2 {
+					return map[string]string{}, fmt.Errorf("'%s' is not formatted as key=value,key1=value1", s)
+				}
```
ideally, this code block is not required as above isValidKeyValuePair fucntion should cover for invalid format, it's just to be safe to have this check.

gandro commented on operator/option/config.go:
```diff
@@ -446,15 +447,15 @@ func (c *OperatorConfig) Populate() {
 	}
 
-	if m := viper.GetStringMapString(IPAMSubnetsTags); len(m) != 0 {
+	if m := command.GetStringMapString(viper.GetViper(), IPAMSubnetsTags); len(m) != 0 {
```
I think we should use `GetStringMapStringE` here, now that we have it, and fail with `log.Fatalf` for any parsing errors. The underlying issue this PR is intending to fix wasn't discovered for almost two years because viper silently swallowed the errors.

gandro commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,85 @@
+var keyValueRegex = regexp.MustCompile(`((\w+)=(\w+,))+((\w+)=(\w+))$`)
+
+// GetStringMapString contains one enhancement to support k1=v2,k2=v2 compared to original
+// implementation of GetStringMapString function
```
Might be worth giving a bit more context here. Could you maybe link to https://github.com/spf13/viper/issues/911 here or in the lint?

gandro reviewed (changes requested):
Awesome, thanks a ton for taking care of this. I believe there might be an issue where we reject single `k=v` pairs. Very neat to see you also adding a lint.

gandro reviewed (approved):
Thanks!

Edit: Marking this for backport to v1.10, as we have users reporting issues with this flag on that version too.

sayboras committed:
cmd: Fix stringer for mapOption

The Set() function for this struct is expecting k=v format,
however, the String() implementation is not returning same format,
which causes some issue while parsing for few fields such as
kvstore-opt, and api-rate-limit.

Relates to cb14db88d7

Signed-off-by: Tam Mach <tam.mach@isovalent.com>

aanm commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// GetStringMapStringE is same as GetStringMapString, but with error
+func GetStringMapStringE(vp *viper.Viper, key string) (map[string]string, error) {
```
Why are we returning an initialized `map[string]string{}` for the error cases?

sayboras replied:
this is to be aligned with upstream function.

https://github.com/spf13/cast/blob/master/caste.go#L863

aanm commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,93 @@
+	if err != nil {
+		var syntaxErr *json.SyntaxError
+		if !errors.As(err, &syntaxErr) {
+			return v, err
```
Is this line covered in the unit tests?

sayboras replied:
good point :100:

ldelossa commented on pkg/command/map_string.go:
```diff
@@ -0,0 +1,93 @@
+
+// isSpace returns true if input is space
+// this is same as the function used in json package
+func isSpace(c byte) bool {
```
@sayboras Before this merges, can you take a look at the "unicode" package. I think some of these methods are redundant, such as: https://pkg.go.dev/unicode#IsSpace

Once over "unicode", "strings" and "bytes" packages to confirm if any of these helper methods have analogies in the stdlib.

sayboras replied:
thanks, I was checking this one from std library, it was a private method though.

https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/encoding/json/scanner.go;l=197-199

sayboras replied:
you are right, it's much simpler with the below

```golang
			// if the input is starting with either '{' or '[', just preserve original json parsing error.
			firstIndex := strings.IndexFunc(s, func(r rune) bool {
				return !unicode.IsSpace(r)
			})
```

aditighag commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -131,7 +131,7 @@ cilium-agent [flags]
       --envoy-log string                                     Path to a separate Envoy log file, if any
       --exclude-local-address strings                        Exclude CIDR from being recognized as local address
-      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities (default map[])
+      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities
```
What do you think of giving an example similar to L20 and L207?

sayboras replied:
good point, let me add example for related field.

aditighag reviewed (changes requested):
LGTM. Have given a suggestion around improving config usability.

joestringer reviewed (approved):
LGTM from my codeowners

aditighag commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -131,7 +131,7 @@ cilium-agent [flags]
       --envoy-log string                                     Path to a separate Envoy log file, if any
       --exclude-local-address strings                        Exclude CIDR from being recognized as local address
-      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities (default map[])
+      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,128=kv-store
```
Why is the k=v repeated in the example?

sayboras replied:
Good spot, it should be 129=kube-dns, i messed up with my local git hook, so the push didn't come through.

sayboras replied:
small diff before force push


```diff
diff --git a/Documentation/cmdref/cilium-agent.md b/Documentation/cmdref/cilium-agent.md
index 45dcb6b942..7733f285c5 100644
--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -131,7 +131,7 @@ cilium-agent [flags]
       --endpoint-status strings                              Enable additional CiliumEndpoint status features (controllers,health,log,policy,state)
       --envoy-log string                                     Path to a separate Envoy log file, if any
       --exclude-local-address strings                        Exclude CIDR from being recognized as local address
-      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,128=kv-store
+      --fixed-identity-mapping map                           Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,129=kube-dns
       --force-local-policy-eval-at-source                    Force policy evaluation of all local communication at the source endpoint (default true)
       --gops-port int                                        Port for gops server to listen on (default 9890)
   -h, --help                                                 help for cilium-agent
diff --git a/daemon/cmd/daemon_main.go b/daemon/cmd/daemon_main.go
index 67199f8cd8..31572ca397 100644
--- a/daemon/cmd/daemon_main.go
+++ b/daemon/cmd/daemon_main.go
@@ -449,7 +449,7 @@ func initializeFlags() {
 	option.BindEnvWithLegacyEnvFallback(option.DisableEnvoyVersionCheck, "CILIUM_DISABLE_ENVOY_BUILD")
 
 	flags.Var(option.NewNamedMapOptions(option.FixedIdentityMapping, &option.Config.FixedIdentityMapping, option.Config.FixedIdentityMappingValidator),
-		option.FixedIdentityMapping, "Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,128=kv-store")
+		option.FixedIdentityMapping, "Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,129=kube-dns")
 	option.BindEnv(option.FixedIdentityMapping)
 
 	flags.Duration(option.IdentityChangeGracePeriod, defaults.IdentityChangeGracePeriod, "Time to wait before using new identity on endpoint identity change")

```

aditighag reviewed (approved):
Approved from contributing
